### PR TITLE
cmetrics: add cmt_expire for expiring metrics.

### PR DESCRIPTION
**Pull request by pwhelan**
Add two new functions, `cmt_map_metrics_expire` for expiring metrics in a `cmt_map` and `cmt_expire` which uses the function to expire all the metrics inside a `cmetrics` context.

This function is useful in general for removing metrics with labels for objects that will disappear, ie: PIDs, etc...

This will also be used in fluent/fluent-bit#7615.

### CODE REVIEW

**Comment by pwhelan** [comment]
With a long running test running under valgrind:

```
==2706702== Memcheck, a memory error detector
==2706702== Copyright (C) 2002-2024, and GNU GPL'd, by Julian Seward et al.
==2706702== Using Valgrind-3.25.1 and LibVEX; rerun with -h for copyright info
==2706702== Command: ./bin/fluent-bit -i mem_metrics -o stdout -f 1
==2706702== Parent PID: 2898109
==2706702== 
==2706702== 
==2706702== HEAP SUMMARY:
==2706702==     in use at exit: 0 bytes in 0 blocks
==2706702==   total heap usage: 27,895,617 allocs, 27,895,617 frees, 4,055,355,463,094 bytes allocated
==2706702== 
==2706702== All heap blocks were freed -- no leaks are possible
==2706702== 
==2706702== For lists of detected and suppressed errors, rerun with: -s
==2706702== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

**Comment by pwhelan** [comment]
I ran an instance of fluent-bit and monitored it from another while running several instances of the following script:

```
while true ; do sleep 1; echo -n . ; done
```

I then graphed out the memory usage overtime which seems to remain consistent:

image

**Comment by piwai** [comment]
Hey @pwhelan , need any help to rebase/update the branch to help get it merged? As you can see in this fluent-bit [issue](https://github.com/fluent/fluent-bit/pull/11760), I'd be extremely interested by this feature, and ended up implementing the same thing but in the wrong place 😅

**Comment by pwhelan** [comment]
> Hey @pwhelan , need any help to rebase/update the branch to help get it merged? As you can see in this fluent-bit [issue](https://github.com/fluent/fluent-bit/pull/11760), I'd be extremely interested by this feature, and ended up implementing the same thing but in the wrong place 😅

gimme a bit, let me see what I can do.

** edit ** rebased. hopefully master has been fixed and tests can pass now.

@piwai tests passed.

@cosmo0920 this PR looks ready to be merged. should I add a unit test or two first?

File: src/cmetrics.c [hunk 1]
```diff
@@ -144,6 +145,42 @@ void cmt_destroy(struct cmt *cmt)
...
+void cmt_expire(struct cmt *cmt, uint64_t expiration)
+{
+    struct cfl_list *tmp;
+    struct cfl_list *head;
+    struct cmt_counter *c;
+    struct cmt_gauge *g;
+    struct cmt_summary *s;
+    struct cmt_histogram *h;
+    struct cmt_untyped *u;
+
```
Comment (cosmo0920) [inline src/cmetrics.c:158]: Do we need to add NULL check before entering an expiration phase?
This is because cmt context pointer access could cause NULL pointer exception when passing NULL in the first argument.

File: tests/expire.c [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+    cmt_destroy(cmt);
+}
+
+TEST_LIST = {
+    {"expire_counter" , test_expire_counter},
+    {"expire_gauge" , test_expire_gauge},
+    {"expire_histogram" , test_expire_histogram},
+    {"expire_summary", test_expire_summary},
+    { 0 }
```
Comment (cosmo0920) [inline tests/expire.c:320]: We need to add untyped case of unit testing here.

**Review by cosmo0920 — CHANGES_REQUESTED** [review]
I found two of the concerns and added as comments.

**Comment by pwhelan** [comment]
> I found two of the concerns and added as comments.

@cosmo0920 I added the NULL parameter check and the test for expiring cmt_untyped.

**Comment by edsiper** [comment]
A few things I think we should settle before merging:

  1. `cmt_map_metrics_expire()` currently removes metrics with `timestamp <= expiration`. The description says “older than”, so I think this should probably be `< expiration`. That also avoids deleting metrics updated during the same collection pass when the caller uses that pass timestamp as the cutoff.

  2. `cmt_expire()` does not walk `exp_histograms`, so the context-wide expiration is incomplete.

  3. `cmt_map_metrics_expire()` returns `int` but always returns `0`. It may be clearer to either return the expired count or make it `void`.

  Could you also add tests for the equal-timestamp boundary and exponential histograms?

**Comment by pwhelan** [comment]
> A few things I think we should settle before merging:

@edsiper all done for now.

** edit ** I'll add an off-by-one test today.

**Review by cosmo0920 — APPROVED** [review]

**Comment by pwhelan** [comment]
I added a test called `test_expire_off_by_one` that tests the boundary by testing expiration at t0-2, t0-1, t0 and t0+1 to expire three labelled counters.

@edsiper anything else before this can get merged?

**Comment by piwai** [comment]
ping @edsiper for review please  :innocent:  (sorry to insist but I'm still very interested by this feature to implement alerting rules on dead process using prometheus process metrics exported by fluent-bit)

**Comment by pwhelan** [comment]
@edsiper or @cosmo0920 is there anything else needed to merge this PR?

**Comment by edsiper** [comment]
I think we are good at this point, would you please do some cleanup in the git history. of the branch ?

**Comment by pwhelan** [comment]
> I think we are good at this point, would you please do some cleanup in the git history. of the branch ?

@edsiper done.

**Comment by edsiper** [comment]
thanks!
